### PR TITLE
Add Gated Embedding Module and Expert Projection Modulation

### PR DESCRIPTION
**Pull request by Naqu6**
Work done:
- Implemented Gated Embedding Module
- Implemented Expert Projection Module

### CODE REVIEW

File: videoretrieval/models/layers/gatedembeddingmodule.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+class GatedEmbeddingModule(tf.keras.layers.Layer):
+    """An implementation of the gating embeding module.
+
+    The gated embedding module takes one tensor (Z_0) of dimension
+    input_dimension as input and returns a tensor of output_dimension.
+
+    The equations that define the gated embedding module are as follows.
+
+    Z_1 = W_1 * Z_0 + B_1
+    Z_2 = Z_1 element wise multiplication by Sigmoid(W_2 * Z_1 + B_2)
```
Comment (sbucak) [inline videoretrieval/models/layers/gatedembeddingmodule.py:30]: nit: If this is Z2 = multiply(Z1, Sigmoid), maybe we can rephrase the expression or use np.multiply or something similar?

File: videoretrieval/models/layers/gatedembeddingmodule.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    Output = Z_2 / norm(Z_2)
+
+    There are four trainable variables, W_1, B_1, W_2, and B_2, which we can
+    represent in linear_layer_one and linear_layer_two.
+
+    Attributes:
+        linear_layer_one: a Dense layer that performs W_1 * Z_0 + B_1.
+        linear_layer_two: a Dense layer that performs W_2 * Z_1 + B_2.
```
Comment (sbucak) [inline videoretrieval/models/layers/gatedembeddingmodule.py:38]: does it also apply sigmoid?

File: videoretrieval/models/layers/gatedembeddingmodule.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            output_dimension, input_shape=(input_dimension,))
+        self.linear_layer_two = tf.keras.layers.Dense(
+            output_dimension, input_shape=(output_dimension,), 
+            activation="sigmoid")
+
+    def call(self, inputs):
+        layer_one_activations = self.linear_layer_one(inputs)
+        layer_two_activations = self.linear_layer_two(layer_one_activations)
+
+        unscaled_activations = layer_one_activations * layer_two_activations
```
Comment (sbucak) [inline videoretrieval/models/layers/gatedembeddingmodule.py:54]: is this supposed to be element-wise product?

File: videoretrieval/models/layers/gatedembeddingmodule.py (hunk 2 above)
Comment (Naqu6) [reply videoretrieval/models/layers/gatedembeddingmodule.py:38]: Yes! Good catch, thank you.

File: videoretrieval/models/layers/gatedembeddingmodule.py (hunk 3 above)
Comment (Naqu6) [reply videoretrieval/models/layers/gatedembeddingmodule.py:54]: Yes

File: videoretrieval/models/layers/gatedembeddingmodule.py (hunk 1 above)
Comment (Naqu6) [reply videoretrieval/models/layers/gatedembeddingmodule.py:30]: Sure! happy to.

**Comment by Naqu6** [comment]
Hey Serhat - made the requested changes. I didn't add a test for GatedEmbeddingModule because it is a trainable layer and I don't think we're planning on having unit tests for those. Happy to figure out a way to add test coverage for that layer though!
